frame/assets/src/lib.rs

```diff
@@ -247,19 +256,594 @@ decl_error! {
+		/// have no existential deposit.
+		/// - `min_balance`: The minimum balance of this new asset that any single account must
+		/// have. If an account's balance is reduced below this, then it collapses to zero.
+		///
+		/// Emits `ForceCreated` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::force_create()]
+		fn force_create(origin,
+			#[compact] id: T::AssetId,
+			owner: <T::Lookup as StaticLookup>::Source,
+			max_zombies: u32,
+			min_balance: T::Balance,
+		) {
+			T::ForceOrigin::ensure_origin(origin)?;
+			let owner = T::Lookup::lookup(owner)?;
+
+			ensure!(!Asset::<T>::contains_key(id), Error::<T>::InUse);
+
+			Asset::<T>::insert(id, AssetDetails {
+				owner: owner.clone(),
+				issuer: owner.clone(),
+				admin: owner.clone(),
+				freezer: owner.clone(),
+				supply: Zero::zero(),
+				deposit: Zero::zero(),
+				max_zombies,
+				min_balance,
+				zombies: Zero::zero(),
+				accounts: Zero::zero(),
+			});
+			Self::deposit_event(RawEvent::ForceCreated(id, owner));
+		}
+
+		/// Destroy a class of fungible assets owned by the sender.
+		///
+		/// The origin must be Signed and the sender must be the owner of the asset `id`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		/// asset.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(z)` where `z` is the number of zombie accounts.
+		#[weight = T::WeightInfo::destroy(*zombies_witness)]
+		fn destroy(origin,
+			#[compact] id: T::AssetId,
+			#[compact] zombies_witness: u32,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Asset::<T>::try_mutate_exists(id, |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+				ensure!(details.owner == origin, Error::<T>::NoPermission);
+				ensure!(details.accounts == details.zombies, Error::<T>::RefsLeft);
+				ensure!(details.zombies <= zombies_witness, Error::<T>::BadWitness);
+				T::Currency::unreserve(&details.owner, details.deposit);
+
+				*maybe_details = None;
+				Account::<T>::remove_prefix(&id);
+				Self::deposit_event(RawEvent::Destroyed(id));
+				Ok(())
+			})
+		}
+
+		/// Destroy a class of fungible assets.
+		///
+		/// The origin must conform to `ForceOrigin`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		/// asset.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::force_destroy(*zombies_witness)]
+		fn force_destroy(origin,
+			#[compact] id: T::AssetId,
+			#[compact] zombies_witness: u32,
+		) -> DispatchResult {
+			T::ForceOrigin::ensure_origin(origin)?;
+
+			Asset::<T>::try_mutate_exists(id, |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+				ensure!(details.accounts == details.zombies, Error::<T>::RefsLeft);
+				ensure!(details.zombies <= zombies_witness, Error::<T>::BadWitness);
```

Comment:
Should the force origin be allowed to ignore these checks?

It would seem hard in general once you distribute a token, to have all of that token returned to kill an asset. Force origin may want to destroy a token, without needing to retrieve all the tokens back from users.

Reply:
No - otherwise we'll leave system account references open with no way of closing them.